operator/controllers/resources/teleport_reconciler.go

```diff
@@ -0,0 +1,100 @@
+package resources
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gravitational/trace"
+	"k8s.io/apimachinery/pkg/api/meta"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+type TeleportKubernetesResource[T types.Resource] interface {
+	kclient.Object
+	ToTeleport() T
+	StatusConditions() *[]v1.Condition
+}
+
+type TeleportResourceReconciler[T types.ResourceWithOrigin, K TeleportKubernetesResource[T]] struct {
+	ResourceBaseReconciler
+	resourceClient TeleportResourceClient[T]
+	kubeResource   K
+}
+
+type TeleportResourceClient[T types.Resource] interface {
+	Get(context.Context, string) (T, error)
+	Create(context.Context, T) error
+	Update(context.Context, T) error
+	Delete(context.Context, string) error
+}
+
+func NewTeleportResourceReconciler[T types.ResourceWithOrigin, K TeleportKubernetesResource[T]](
+	client kclient.Client,
+	resourceClient TeleportResourceClient[T],
+	kubeResource K) *TeleportResourceReconciler[T, K] {
+
+	reconciler := &TeleportResourceReconciler[T, K]{
+		ResourceBaseReconciler: ResourceBaseReconciler{Client: client},
+		resourceClient:         resourceClient,
+		kubeResource:           kubeResource,
+	}
+	reconciler.ResourceBaseReconciler.UpsertExternal = reconciler.Upsert
+	reconciler.ResourceBaseReconciler.DeleteExternal = reconciler.Delete
+	return reconciler
+}
+
+func (r TeleportResourceReconciler[T, K]) Upsert(ctx context.Context, obj kclient.Object) error {
+	k8sResource, ok := obj.(K)
+	if !ok {
+		return fmt.Errorf("failed to convert Object into resource object: %T", obj)
+	}
+	teleportResource := k8sResource.ToTeleport()
+
+	existingResource, err := r.resourceClient.Get(ctx, teleportResource.GetName())
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	exists := !trace.IsNotFound(err)
```

Comment:
Is this correct check? If `err` is not "not found" then it would've been returned just above, no? Wouldn't more straightforward check be just `exists := existingResource != nil`?

Reply:
`existingResource` is of type `T`, T is a pointer but, it technically could be something else.
The compiler will not let met compare `T` and `untyped nil`.


this check is basically checking between `err == nil` and `trace.IsNotFound(err)`
```
exists := !trace.IsNotFound(err)
```
is equivalent to
```
exists := err != nil
```

I think both are confusing, I'll change it to `err != nil`